File: python/sol_cuda/PmlEz.py
# -*- coding: utf-8 -*-
"""
PmlEz.py (CUDA)
"""

import math
from numba import cuda

def calEz(block1d,
    Nx, Ny, Ez, Hx, Hy, Ezx, Ezy, RXn, RYn,
    iPmlEz, gPmlXn, gPmlYn, rPmlE, rPml, pml_l,
    Ni, Nj, Nk, N0):

    numEz = iPmlEz.shape[0]
    grid = math.ceil(numEz / block1d)

    _calcEz_gpu[grid, block1d](
        Nx, Ny, Ez, Hx, Hy, Ezx, Ezy, RXn, RYn,
        numEz, iPmlEz, gPmlXn, gPmlYn, rPmlE, rPml, pml_l,
        Ni, Nj, Nk, N0)

# (private)
@cuda.jit(cache=True)
def _calcEz_gpu(
    Nx, Ny, Ez, Hx, Hy, Ezx, Ezy, RXn, RYn,
    numEz, iPmlEz, gPmlXn, gPmlYn, rPmlE, rPml, pml_l,
    Ni, Nj, Nk, N0):

    num = cuda.grid(1)

    if num < numEz:
        i, j, k, m = iPmlEz[num]

        n   = Ni * (i    ) + Nj * (j    ) + Nk * (k    ) + N0
        ni1 = Ni * (i - 1) + Nj * (j    ) + Nk * (k    ) + N0
        nj1 = Ni * (i    ) + Nj * (j - 1) + Nk * (k    ) + N0

        ic = min(max(i, 0), Nx    )
        dhy = Hy[n] - Hy[ni1]
        rx = RXn[ic] * rPmlE[m]
        Ezx[num] = (Ezx[num] + (rx * dhy)) / (1 + (gPmlXn[i + pml_l] * rPml[m]))

        jc = min(max(j, 0), Ny    )
        dhx = Hx[n] - Hx[nj1]
        ry = RYn[jc] * rPmlE[m]
        Ezy[num] = (Ezy[num] - (ry * dhx)) / (1 + (gPmlYn[j + pml_l] * rPml[m]))

        Ez[n] = Ezx[num] + Ezy[num]

File: sol_sycl/pmlEz.c
/*
pmlEz.c (OpenMP)

PML for Ez
*/

#include "ofd.h"

#ifdef _ONEAPI
#undef C	// C is used for (2.99792458e8) but <CL/sycl.hpp> refuses it
#include "ofd_dpcpp.h"
#endif

void pmlEz(void)
{
	const int lx = cPML.l;
	const int ly = cPML.l;

#ifndef _ONEAPI
	int64_t n;
#ifdef _OPENMP
#pragma omp parallel for
#endif
#ifdef __CLANG_FUJITSU
#pragma clang loop vectorize(assume_safety)
#endif
	for (n = 0; n < numPmlEz; n++) {
		const int  i = fPmlEz[n].i;
		const int  j = fPmlEz[n].j;
		const int  k = fPmlEz[n].k;
		const id_t m = fPmlEz[n].m;

		const real_t dhy = HY(i, j, k) - HY(i - 1, j, k);
		const real_t rx = RXn[MIN(MAX(i, 0), Nx    )] * rPmlE[m];
		Ezx[n] = (Ezx[n] + (rx * dhy)) / (1 + (gPmlXn[i + lx] * rPml[m]));

		const real_t dhx = HX(i, j, k) - HX(i, j - 1, k);
		const real_t ry = RYn[MIN(MAX(j, 0), Ny    )] * rPmlE[m];
		Ezy[n] = (Ezy[n] - (ry * dhx)) / (1 + (gPmlYn[j + ly] * rPml[m]));

		EZ(i, j, k) = Ezx[n] + Ezy[n];
	}
#else //_ONEAPI
	const int pmlBlock = 256; 
	sycl::range<1> updateBlock = sycl::range<1>(pmlBlock);
	sycl::range<1> grid(CEIL(numPmlEz, pmlBlock));
	sycl::range<1> all_grid = grid * updateBlock;

	myQ.submit([&](sycl::handler& hndl) {
		auto numPmlEz = ::numPmlEz;
		auto fPmlEz = ::fPmlEz;
		auto Hx = ::Hx;
		auto Hy = ::Hy;
		auto RXn = :: RXn;
		auto RYn = :: RYn;
		auto rPmlE = ::rPmlE;
		auto rPml = ::rPml;
		auto gPmlXn = ::gPmlXn;
		auto gPmlYn = ::gPmlYn;

		auto Ezx = ::Ezx;
		auto Ezy = ::Ezy;
		auto Ez = ::Ez;
		auto Nx = ::Nx;
		auto Ny = ::Ny;
		auto Ni = ::Ni;
		auto Nj = ::Nj;
		auto Nk = ::Nk;
		auto N0 = ::N0;


		hndl.parallel_for(
			sycl::nd_range<1>(all_grid, updateBlock),
			[=](sycl::nd_item<1> idx) {
				const int64_t n = idx.get_global_id(0);
				if (n < numPmlEz){
		const int  i = fPmlEz[n].i;
		const int  j = fPmlEz[n].j;
		const int  k = fPmlEz[n].k;
		const id_t m = fPmlEz[n].m;

		const real_t dhy = HY(i, j, k) - HY(i - 1, j, k);
		const real_t rx = RXn[MIN(MAX(i, 0), Nx    )] * rPmlE[m];
		Ezx[n] = (Ezx[n] + (rx * dhy)) / (1 + (gPmlXn[i + lx] * rPml[m]));

		const real_t dhx = HX(i, j, k) - HX(i, j - 1, k);
		const real_t ry = RYn[MIN(MAX(j, 0), Ny    )] * rPmlE[m];
		Ezy[n] = (Ezy[n] - (ry * dhx)) / (1 + (gPmlYn[j + ly] * rPml[m]));

		EZ(i, j, k) = Ezx[n] + Ezy[n];
				}
			});
		});
	myQ.wait();
#endif // _ONEAPI
}
